gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java

```diff
@@ -219,6 +223,178 @@ public void createBucket(String bucketName, CreateBucketOptions options) throws
+                try {
+                  canIgnoreException = canIgnoreExceptionForEmptyObject(se, resourceId, options);
+                } catch (Exception e) {
+                  // Make sure to catch Exception instead of only StorageException so that we can
+                  // correctly wrap other such exceptions and propagate them out cleanly inside
+                  // innerExceptions.
+                  innerExceptions.add(
+                      new IOException(
+                          "Error re-fetching after rate-limit error: " + resourceId, e));
+                }
+                if (canIgnoreException) {
+                  logger.atInfo().log(
+                      "Ignoring exception of type %s; verified object already exists with desired"
+                          + " state.",
+                      se.getClass().getSimpleName());
+                  logger.atFine().withCause(se).log(
+                      "Ignored exception while creating empty object");
+                } else {
+                  innerExceptions.add(new IOException("Error inserting " + resourceId, se));
+                }
+              }
+              return null;
+            },
+            null);
+      }
+    } finally {
+      executor.shutdown();
+    }
+
+    if (!innerExceptions.isEmpty()) {
+      throw GoogleCloudStorageExceptions.createCompositeException(innerExceptions);
+    }
+  }
+
+  private void createEmptyObjectInternal(
+      StorageResourceId resourceId, CreateObjectOptions createObjectOptions) {
+    Map<String, String> rewrittenMetadata = encodeMetadata(createObjectOptions.getMetadata());
+
+    List<BlobTargetOption> blobTargetOptions = new ArrayList<>();
+    blobTargetOptions.add(BlobTargetOption.disableGzipContent());
+    if (resourceId.hasGenerationId()) {
+      blobTargetOptions.add(BlobTargetOption.generationMatch(resourceId.getGenerationId()));
+    } else if (resourceId.isDirectory() || !createObjectOptions.isOverwriteExisting()) {
+      blobTargetOptions.add(BlobTargetOption.generationMatch(0L));
+    }
+
+    if (storageOptions.getEncryptionKey() != null) {
+      blobTargetOptions.add(
+          BlobTargetOption.encryptionKey(storageOptions.getEncryptionKey().value()));
+    }
+
+    storage.create(
+        BlobInfo.newBuilder(BlobId.of(resourceId.getBucketName(), resourceId.getObjectName()))
+            .setMetadata(rewrittenMetadata)
+            .setContentEncoding(createObjectOptions.getContentEncoding())
+            .setContentType(createObjectOptions.getContentType())
+            .build(),
+        blobTargetOptions.toArray(BlobTargetOption[]::new));
+  }
+
+  /**
+   * Helper to check whether an empty object already exists with the expected metadata specified in
+   * {@code options}, to be used to determine whether it's safe to ignore an exception that was
+   * thrown when trying to create the object, {@code exceptionOnCreate}.
+   */
+  private boolean canIgnoreExceptionForEmptyObject(
```

Comment:
We are doing a lot of code duplication. Is there a way to avoid it?

Reply:
I believe, we can avoid this duplication by refactoring the GoogleCloudStorageImpl at bit. However, given that we might deprecate GoogleCloudStorageImpl in future. I dont see merit in investing in this re-factoring at the moment. Let me know if you have any other thoughts.